### ISSUE TITLE
Refactor request - Don't use cryptocurrency libraries

### ISSUE THREAD

**Issue by craig-seeman** [chaos-mesh/chaos-mesh#2556]
## Feature Request

I was looking at dependencies and noticed chaos-mesh uses the ethereum library for json rpc communication in the iochaos_server. I am wondering if someone could refactor the logic within to be similar to httpschaos_server.go. It appears that httpchaos_server is using http calls over STDIO, and I am wondering if the same method could be used (with a timeout) in iochaos, instead of loading this complex ethereum library for something so simple.

Previous discussion in PR https://github.com/chaos-mesh/chaos-mesh/pull/1545#discussion_r603053771

**Comment by craig-seeman** [comment]
Actually looking at this, :( it will probably require a refactor on toda.. I see that https://github.com/chaos-mesh/rs-tproxy is using http calls to do live reload and such, I wonder how hard it would be to refactor that as well.. though it seems like quite some work..

**Comment by YangKeao** [comment]
@Hexilee How do you think about it? I forgot the reason why we choose jsonrpc but not http for toda 😢

**Comment by Hexilee** [comment]
> @Hexilee How do you think about it? I forgot the reason why we choose jsonrpc but not http for toda 😢

I prefer HTTP over stdio, but at that time we were refactoring IOChaos, @AsterNighT thought HTTP is much more complex than jsonrpc.

**Comment by Hexilee** [comment]
> @Hexilee How do you think about it? I forgot the reason why we choose jsonrpc but not http for toda 😢

I prefer http over stdio, but at that time we were refactoring IOChaos, @AsterNighT thought http is much more complex than jsonrpc.

**Comment by Colstuwjx** [comment]
How about make a chaos-mesh version codec for stdio? We could directly use go official version [jsonrpc](https://pkg.go.dev/net/rpc/jsonrpc), and make a stdio conn codec for it, like go-ethereum [did](https://github.com/ethereum/go-ethereum/blob/master/rpc/stdio.go#L36).

**Comment by YangKeao** [comment]
> How about make a chaos-mesh version codec for stdio? We could directly use go official version [jsonrpc](https://pkg.go.dev/net/rpc/jsonrpc), and make a stdio conn codec for it, like go-ethereum [did](https://github.com/ethereum/go-ethereum/blob/master/rpc/stdio.go#L36).

Go official jsonrpc only supports jsonrpc 1.0. On the other side, no rust crates supports jsonrpc 1.0, that's why we are using the jsonrpc from ethereum now.

**Comment by craig-seeman** [comment]
An interesting idea could be to use grpc over stdio. Since much of the internals of chaos-mesh already utilize grpc. Through some quick searching, it seems that there may be some support for this in rust and other languages (over stdio); meaning it could be a language agnostic decision.

However I feel bad suggesting all these things to redo so much work that @AsterNighT and you all have put into these current things.. 

It was just alarming to me when I saw ethereum dependency (perhaps I worry too much 😅), I would not want others to see it too and have them require investigation into why it is used - which is a shame, because it is working well in this project.

**Comment by craig-seeman** [comment]
An interesting idea could be to use grpc over stdio. Since much of the internals of chaos-mesh already utilize grpc. Through some quick searching, it seems that there may be some support for this in rust and other languages (over stdio); meaning it could be a language agnostic decision.

However I feel bad suggesting all these things to redo so much work that @AsterNighT and you all have put into these current things..

**Comment by Hexilee** [comment]
> An interesting idea could be to use grpc over stdio. Since much of the internals of chaos-mesh already utilize grpc. Through some quick searching, it seems that there _may_ be some support for this in rust and other languages (over stdio); meaning it could be a language agnostic decision.
> 
> However I feel bad suggesting all these things to redo so much work that @AsterNighT and you all have put into these current things..

I think http just works well in this case, and gRPC is too complex (which is based on http).  We can reimplement the stdio communication as same as HTTPChaos.

**Comment by craig-seeman** [comment]
> Moreover, a simple self-defined protocol is also acceptable as toda has not enabled configuration hot-reload. We just need to send the configuration to toda once, then monitor its status.

I was wondering about that, I really don't mind the jsonrpc workflow, but perhaps it would be very simple to just implement basic json client for the time being without a full entire standard and jsonrpc client 

(Basically just send and parse json over the stdio ourselves). In the daemonset as you said, we could just send manual encoded message such as this https://github.com/chaos-mesh/toda/blob/master/config-examples/latency-example.json

I was attempting this last evening to test that, however the pings were not returning as I was expecting, which was very confusing to me. (But I was probably attempting it incorrectly, lol..)

Request --->
```
{"jsonrpc":"2.0","method":"get_status","params":["ping"],"id":3}
```
<--- Response (I was expecting "Ok" in result)
```
{"jsonrpc":"2.0","result":"source: /app/test, target: /app/__chaosfs__test__","id":3}
```

**Comment by YangKeao** [comment]
> `{"jsonrpc":"2.0","result":"source: /app/test, target: /app/__chaosfs__test__","id":3}`

This is an error message when it tries to move the mount point from `/app` to `/app/__chaosfs_test__` but failed. (However, the error message is not encoded well. The errno and detail message lost 😢 )

From experience, the most common problem is that the `/app` doesn't exist, or the parent mount of `/app` has propagation type MS_SHARED.

**Comment by craig-seeman** [comment]
Now that I'm thinking on it, perhaps this could be possibly an RFC to allow for future ease of adding similar attack functionality?

**Comment by Hexilee** [comment]
I think it's a detail about special implementation, which depends on programming languages and platforms. An RFC may be unnecessary, I would like a detailed tutorial or SDKs between Golang and Rust applications.

**Comment by craig-seeman** [comment]
Makes sense. It would just be some ways to say 'if you create a separate binary or script and want to communicate to it, stdio http is the standard built into chaos-mesh'

The reason I was wondering this is because it would be interesting approach to have ALL attacks happen in the same way the toda/tproxy is written. 

This would result in the daemonset launching each attack that is self-contained within their own binary / script. If this is done, it could help decouple the 'what to attack' vs 'how to attack' and simplify the code within the base repo a bit (especially the daemonset).

**Comment by craig-seeman** [comment]
Makes sense. It would just be some ways to say 'if you create a separate binary or script (any language) and want to communicate to it, stdio http (or whatever is chose) is the standard built into chaos-mesh'

**Comment by STRRL** [comment]
Why not UNIX domain socket? 🤔

**Comment by YangKeao** [comment]
> Why not UNIX domain socket? 🤔

`toda` needs to communicate cross the mnt namespace, which is not possible (AFAIK) through unix domain socket.

**Comment by Colstuwjx** [comment]
> Why not UNIX domain socket? 🤔

BTW, It's not about how to do ipc communication, the problem is that, which communication protocol is the best choice.

For now, IOChaos server forked a sub process to run toda, and toda is running into the container's mnt namespace, IOChaos server communicates with toda via the toda process' stdio since pid namespace isn't isolated. I think it's a pretty good solution.

The problem is, in golang, we don't have an official and easy-to-use lightweight solution for rpc 2.0, even the rpc 1.0 is [problematic](https://github.com/golang/go/issues/16844).

**Comment by STRRL** [comment]
> Why not UNIX domain socket? thinking

Why not gRPC? 🙈

**Comment by Hexilee** [comment]
> > Why not UNIX domain socket? thinking
> 
> Why not gRPC? 🙈

gRPC is based on HTTP, and HTTP meet all requirements in this case.

**Comment by STRRL** [comment]
Hi guys, I just realize that there is a similar situation on https://github.com/containernetworking/cni/issues/785.

CNI v1 used stdio for communication, and they might be going to use gRPC in the future(like CSI). 🤣

And use UnixDomainSocket might bring some trouble with different mount namespace, but I am sure for that we could provide a  useful tool like a "pipe" that could bridge two unix domain socket files in different mount namespace(like veth pair in two network namespaces).

**Comment by YangKeao** [comment]
@STRRL
> but I am sure for that we could provide a useful tool like a "pipe" that could bridge two unix domain socket files in different mount namespace

Yes. I can tweak the `nsexec` to let the subprocess inherit some file descriptor.

**Comment by STRRL** [comment]
> What's more, the nsexec can also inherit a pipe from the chaos-daemon, which means we can use an anonymous pipe rather than unix socket.

That's interesting, what would you like to use instead of UNIX socket domain for working as "pipe"?

**Comment by YangKeao** [comment]
>  what would you like to use instead of UNIX socket domain for working as "pipe"?

@STRRL Anonymous [pipe](https://man7.org/linux/man-pages/man2/pipe.2.html), but I'm not sure whether it will work. I will try to migrate the jsonrpc over stdio to jsonrpc over pipe (and remove the blocking buffer in bpm), then consider other rpc (like grpc).

**Comment by YangKeao** [comment]
@STRRL I have drafted a proposal about using abstract unix socket to communicate across the namespaces. https://github.com/chaos-mesh/rfcs/pull/34
